File: problems/Accounts Merge/accountsMerge.cs
public class Solution {
    public IList<IList<string>> AccountsMerge(IList<IList<string>> accounts) {
        var result = new Dictionary<int, IList<string>>();
        var dsu = new Dsu(10001);
        var emailToName = new Dictionary<string, string>();
        var emailToId = new Dictionary<string, int>();
        var id = 0;

        foreach (var account in accounts) {
            var name = account[0];

            for (var i = 1; account.Count > i; ++i) {
                var email = account[i];

                if (!emailToName.ContainsKey(email)) {
                    emailToName.Add(email, name);
                }

                if (!emailToId.ContainsKey(email)) {
                    emailToId.Add(email, id++);
                }

                dsu.union(emailToId[account[1]], emailToId[email]);
            }
        }

        foreach (var email in emailToName.Keys) {
            var idx = dsu.find(emailToId[email]);

            if (result.ContainsKey(idx)) {
                result[idx].Add(email);
            } else {
                result.Add(idx, new List<string> { email });
            }
        }

        foreach (var idx in result.Keys) {
            ((List<string>)result[idx]).Sort(
                (x, y) => string.CompareOrdinal(x, y)
            );
            result[idx].Insert(0, emailToName[result[idx][0]]);
        }

        return result.Values.ToList();
    }

    private class Dsu {
        private readonly int[] _parents;
        private readonly int[] _ranks;

        public Dsu(int n) {
            _parents = new int[n];
            _ranks = new int[n];

            for (var i = 0; n > i; ++i) {
                _parents[i] = i;
            }
        }

        public int find(int x) {
            if (x != _parents[x]) {
                x = find(_parents[x]);
            }

            return _parents[x];
        }

        public bool union(int x, int y) {
            var px = find(x);
            var py = find(y);

            if (px == py) {
                return false;
            }

            if (_ranks[px] > _ranks[py]) {
                _parents[py] = px;
                ++_ranks[px];
            } else {
                _parents[px] = py;
                ++_ranks[py];
            }

            return true;
        }
    }
}

File: problems/Accounts Merge/accountsMerge.java
class Solution {
    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        List<List<String>> result = new ArrayList<>();
        Dsu dsu = new Dsu(10000);
        HashMap<Integer, List<String>> idToEmailsMap = new HashMap<>();
        HashMap<String, String> emailToNameMap = new HashMap<>();
        HashMap<String, Integer> emailToIdMap = new HashMap<>();
        int id = 0;

        for (List<String> account : accounts) {
            String name = account.get(0);

            for (int idx = 1; account.size() > idx; ++idx) {
                String email = account.get(idx);

                if (!emailToNameMap.containsKey(email)) {
                    emailToNameMap.put(email, name);
                }

                if (!emailToIdMap.containsKey(email)) {
                    emailToIdMap.put(email, id);
                    ++id;
                }

                dsu.union(emailToIdMap.get(account.get(1)), emailToIdMap.get(email));
            }
        }

        for (Map.Entry<String, Integer> emailAndId : emailToIdMap.entrySet()) {
            int parentId = dsu.find(emailAndId.getValue());

            if (!idToEmailsMap.containsKey(parentId)) {
                idToEmailsMap.put(parentId, new ArrayList<String>());
            }

            idToEmailsMap.get(parentId).add(emailAndId.getKey());
        }

        for (int parentId : idToEmailsMap.keySet()) {
            String name = emailToNameMap.get(idToEmailsMap.get(parentId).get(0));

            Collections.sort(idToEmailsMap.get(parentId));
            idToEmailsMap.get(parentId).add(0, name);
        }

        return new ArrayList(idToEmailsMap.values());
    }

    private class Dsu {
        int[] parents;
        int[] ranks;

        public Dsu(int n) {
            parents = new int[n];
            ranks = new int[n];

            for (int idx = 0; n > idx; ++idx) {
                parents[idx] = idx;
            }
        }

        public int find(int a) {
            if (parents[a] != a) {
                parents[a] = find(parents[a]);
            }

            return parents[a];
        }

        public void union(int a, int b) {
            int parentA = find(a);
            int parentB = find(b);

            if (parentA != parentB) {
                if (ranks[parentA] > ranks[parentB]) {
                    parents[parentB] = parentA;
                    ++ranks[parentA];
                } else {
                    parents[parentA] = parentB;
                    ++ranks[parentB];
                }
            }
        }
    }
}
